### PR TITLE
Changing emailaddres from recipientAddress to recipientAddresses

### PR DESCRIPTION
**Pull request by AyhamAlzoubi**
This PR enable sending emails to multi recipient addresses.   
Some refactoring to remove the public setters.

### CODE REVIEW

**Comment by mohdhallal** [comment]
Do you mean `recipientsAddresses` ??

File: examples/email-custom-handler.php [hunk 1]
```diff
@@ -15,7 +15,7 @@ class SimpleEmailHandler extends EmailHandler
 {    
     public function handle(NotificationInterface $notification)
     {
-        mail($notification->getRecipientAddress(), $notification->subject, $notification->body);
+        mail($notification->getRecipientAddresses(), $notification->subject, $notification->body);
```
Comment (odino) [inline examples/email-custom-handler.php]: does `mail` support multiple addresses??

Comment (AyhamAlzoubi) [inline examples/email-custom-handler.php]: Yes, I just checked it now, you pass a string with comma separated email addresses.
I'm going to modify it.

File: src/Namshi/Notificator/Notification/Email/EmailNotification.php [hunk 2]
```diff
@@ -9,33 +9,38 @@
...
      * 
-     * @param string $recipientAddress
+     * @param array $recipientAddress
      * @param array $parameters
      */
-    public function __construct($recipientAddress, array $parameters = array())
+    public function __construct(array $recipientAddresses, array $parameters = array())
     {
-        $this->setRecipientAddress($recipientAddress);
+        $this->setRecipientAddresses($recipientAddresses);
```
Comment (cirpo) [inline src/Namshi/Notificator/Notification/Email/EmailNotification.php]: let's access the instance variable directly and lets remove the setter. It seems that we are using it just in the constructors

File: spec/Namshi/Notificator/Notification/Email/Emailvision/EmailvisionNotificationSpec.php [hunk 3]
```diff
@@ -8,7 +8,7 @@ class EmailvisionNotificationSpec extends ObjectBehaviorComplete
 {
     function let()
     {
-        $this->beConstructedWith('template', 'recipient', []);
+        $this->beConstructedWith('template', ['recipient'], []);
```
Comment (hossam-fares) [inline spec/Namshi/Notificator/Notification/Email/Emailvision/EmailvisionNotificationSpec.php]: Do emailvision accept multiple recipients?

File: spec/Namshi/Notificator/Notification/Handler/HipChatSpec.php [hunk 4]
```diff
@@ -26,7 +26,7 @@ function it_is_initializable()
     function it_should_handle_hipchat_notifications_only()
     {
         $hipChatNotification = new HipChatNotification('a', 'b', []);
-        $otherNotification = new EmailNotification('a', []);
+        $otherNotification = new EmailNotification(['a'], []);
         if (!$this->getWrappedObject()->shouldHandle($hipChatNotification)) {
```
Comment (hossam-fares) [inline spec/Namshi/Notificator/Notification/Handler/HipChatSpec.php]: NL!

File: spec/Namshi/Notificator/Notification/Handler/NotifySendSpec.php [hunk 5]
```diff
@@ -29,7 +29,7 @@ function it_is_initializable()
     function it_should_handle_notifysend_notifications_only($finder)
     {
         $notification = new NotifySendNotification('message', []);
-        $otherNotification = new EmailNotification('recipient', []);
+        $otherNotification = new EmailNotification(['recipient'], []);
         $finder->find(NotifySend::NOTIFY_SEND_COMMAND)->willReturn(true);
         if (!$this->getWrappedObject()->shouldHandle($notification)) {
```
Comment (hossam-fares) [inline spec/Namshi/Notificator/Notification/Handler/NotifySendSpec.php]: NL!

File: spec/Namshi/Notificator/Notification/Handler/RabbitMQSpec.php [hunk 6]
```diff
@@ -27,7 +27,7 @@ function it_is_initializable()
     function it_should_handle_rabbitmq_notifications_only()
     {
         $notification = new RabbitMQNotification('message', []);
-        $otherNotification = new EmailNotification('recipient', []);
+        $otherNotification = new EmailNotification(['recipient'], []);
```
Comment (hossam-fares) [inline spec/Namshi/Notificator/Notification/Handler/RabbitMQSpec.php]: NL!

**Comment by hossam-fares** [comment]
This implementation will break emailvesion notifications since it accept only one recipient.

**Comment by AyhamAlzoubi** [comment]
@odino @mohdhallal @hossam-fares 
This PR is ready and now it will work for Emailvision and mail.

File: src/Namshi/Notificator/Notification/Email/SwiftMailer/SwiftMailerNotification.php [hunk 7]
```diff
@@ -18,26 +18,15 @@ class SwiftMailerNotification extends EmailNotification implements NotificationI
...
+     * @param array         $parameters
      */
     public function __construct($emailTemplate, $recipientAddress, array $parameters = array())
     {
-        $this->setEmailTemplate($emailTemplate);
-        $this->setRecipientAddress($recipientAddress);
-        $this->setParameters($parameters);
-    }
-
+        parent::__construct($recipientAddress, $parameters);
```
Comment (odino) [inline src/Namshi/Notificator/Notification/Email/SwiftMailer/SwiftMailerNotification.php]: do we wanna pass only one recipient address? if we dont then we could think of a better way to name this function

File: src/Namshi/Notificator/Notification/Email/EmailNotification.php [hunk 8]
```diff
@@ -9,33 +9,30 @@
...
+     * Constructor.
+     * 
+     * @param array|string $recipientAddress
+     * @param array $parameters
      */
-    public function getRecipientAddress()
+    public function __construct($recipientAddresses, array $parameters = array())
     {
-        return $this->recipientAddress;
+        $this->recipientAddresses = is_array($recipientAddresses) ? $recipientAddresses : [$recipientAddresses];
```
Comment (odino) [inline src/Namshi/Notificator/Notification/Email/EmailNotification.php]: btw [this](https://github.com/symfony/Finder/blob/master/Finder.php#L667-L684) is how symfony does it

File: src/Namshi/Notificator/Notification/Email/SwiftMailer/SwiftMailerNotification.php (hunk 7 above)
Comment (AyhamAlzoubi) [inline src/Namshi/Notificator/Notification/Email/SwiftMailer/SwiftMailerNotification.php]: No, Its a string or array of addresses.
I will change the variable name to `$recipientAddresses`

**Comment by AyhamAlzoubi** [comment]
@odino it's ready

**Comment by odino** [comment]
What about just adding an example in the README?

File: spec/Namshi/Notificator/Notification/Email/EmailNotificationSpec.php [hunk 9]
```diff
@@ -17,4 +17,18 @@ function it_is_initializable()
...
+        $this->getRecipientAddresses()->shouldReturn(['recipient']);
+    }
+
+    function it_should_handle_array_of_recipient()
+    {
+        $this->beConstructedWith(['recipient', 'recipient2', 'recipient3']);
+
+        $this->getRecipientAddresses()->shouldReturn(['recipient', 'recipient2', 'recipient3']);
+    }
 }
```
Comment (cirpo) [inline spec/Namshi/Notificator/Notification/Email/EmailNotificationSpec.php]: BL missing

File: src/Namshi/Notificator/Notification/Email/SwiftMailer/SwiftMailerNotification.php [hunk 10]
```diff
@@ -18,26 +18,15 @@ class SwiftMailerNotification extends EmailNotification implements NotificationI
...
 
-    /**
-     * Sets the email template that should be used for this notification.
-     *
-     * @param string $emailTemplate
-     */
-    public function setEmailTemplate($emailTemplate)
-    {
-        $this->emailTemplate = $emailTemplate;
+        $this->emailTemplate      = $emailTemplate;
```
Comment (cirpo) [inline src/Namshi/Notificator/Notification/Email/SwiftMailer/SwiftMailerNotification.php]: too many spaces

File: src/Namshi/Notificator/Notification/Handler/HipChat.php [hunk 11]
```diff
@@ -58,14 +58,4 @@ public function getHipchat()
...
-    /**
-     * Sets the HipChat instance associated with this handler.
-     * 
-     * @param \HipChat\HipChat $hipchat
-     */
-    public function setHipchat(HipChatClient $hipchat)
-    {
-        $this->hipchat = $hipchat;
-    }
 }
```
Comment (cirpo) [inline src/Namshi/Notificator/Notification/Handler/HipChat.php]: BL missing

File: src/Namshi/Notificator/Notification/Handler/RabbitMQ.php [hunk 12]
```diff
@@ -46,9 +46,4 @@ public function getPublisher()
     {
         return $this->publisher;
     }
-
-    public function setPublisher(AMQPChannel $publisher)
-    {
-        $this->publisher = $publisher;
-    }
 }
```
Comment (cirpo) [inline src/Namshi/Notificator/Notification/Handler/RabbitMQ.php]: BL missing

File: src/Namshi/Notificator/NotificationInterface.php [hunk 13]
```diff
@@ -25,12 +25,4 @@ function getParameters();
...
     public function getMessage();
-    
-    /**
-     * Sets an array of parameters that are sent along with this
-     * notification.
-     * They are usually used by handlers to build the actual notification (ie. 
-     * dynamic content in emails or smses).
-     */
-    function setParameters(array $parameters);
 }
```
Comment (cirpo) [inline src/Namshi/Notificator/NotificationInterface.php]: BL missing

**Comment by odino** [comment]
btw @cirpo / @AyhamAlzoubi what about this one? @hossam-fares made a comment on the emailvision handler, not sure we cought that

**Comment by AyhamAlzoubi** [comment]
Yes we already changed it.

**Pull request merged**